lib/membrane_rtc_engine/endpoints/hls_endpoint.ex

```diff
@@ -166,6 +168,45 @@ if Enum.all?(
       {:ok, state}
     end
 
+    @impl true
+    def handle_notification(
+          {:variant_switched, _new_variant},
+          {:track_receiver, track_id},
+          _ctx,
+          state
+        ) do
+      track = Map.fetch!(state.tracks, track_id)
+
+      # this can be simplified after MC-72
+      # instead of checking whether timer was
+      # created or not, we can create timer
+      # with `:no_interval` option in `handle_pad_added`
+      # and always stop and start it here
+      {actions, state} =
+        cond do
+          track.type == :audio ->
+            {[], state}
+
+          MapSet.member?(state.started_tracks, track_id) ->
+            actions = [
+              {:stop_timer, {:request_keyframe, track_id}},
+              {:start_timer, {{:request_keyframe, track_id}, state.target_segment_duration}}
+            ]
+
+            {actions, state}
+
+          true ->
+            actions = [
+              {:start_timer, {{:request_keyframe, track_id}, state.target_segment_duration}}
+            ]
+
+            state = Map.update!(state, :started_tracks, &MapSet.put(&1, track_id))
+            {actions, state}
+        end
+
+      {{:ok, actions}, state}
+    end
```

Comment:
Sorry, but this code smells like spaghetti. To analyze it I'd have to jump back and forth between this endpoint and  `TrackReceiver`. Even assuming MC-72 is fixed, IMO this is overly complicated without enough justification.

Following up on arguments from https://github.com/membraneframework/membrane_rtc_engine/pull/169#discussion_r981296393

> I don't want it to be confused with TrackSender behavior. Whenever TrackSender receives keyframe request it starts its internal timer and requests keyframe every 500ms. Adding keyframe_request_interval option to TrackReceiver might make it less clear what is the difference.

I don't see how that would be more confusing than the current solution - you still have both Sender and Receiver sending keyframe requests and two timers controlling it. Right now it may be harder to find them both, but once you do, the situation is similar with the exception of determining when the requests are sent, as that logic is split between Endpoint and Receiver.

To reduce the risk of confusion I'd ensure the naming in `TrackSender` is proper and highlights the difference, e.g. sender should be using `keyframe_request_repeat_interval` instead of `keyframe_request_interval` or use other keywords depending on the context, e.g. "throttle_interval` or `debounce_interval`.

Another reason I think it would be better to put it there is repetition - imagine creating `dash_endpoint`. It would require repeating that logic and I suspect more endpoints might need it.

Reply:
I think that the code isn't so difficult. If we have video track and we haven't started timer yet, we are starting it. If we have started timer, we are restarting it.

> Another reason I think it would be better to put it there is repetition - imagine creating dash_endpoint. It would require repeating that logic and I suspect more endpoints might need it.

Current API of TrackReceiver is more flexible as everyone can request keyframe both in regular and irregular time periods. 
The same situation is with Process.send_after. Everyone implements periodic messages on its own.


With MC-72 fixed we would have 

```elixir
    @impl true
    def handle_notification(
          {:variant_switched, _new_variant},
          {:track_receiver, track_id},
          _ctx,
          state
        ) do
      track = Map.fetch!(state.tracks, track_id)

      actions =
        if track.type == :audio do
            []
        else
            [
              {:stop_timer, {:request_keyframe, track_id}},
              {:start_timer, {{:request_keyframe, track_id}, state.target_segment_duration}}
            ]
        end
            
      {{:ok, actions}, state}
    end
```

Reply:
There indeed might not be use cases for sending irregular keyframe requests from the bin. I also agree that we should encourage users to send keyframe requests via events so it might indeed be clearer to allow only for periodic keyframe requests and if someone wants irregular ones they can be sent via events.

However I still think that the additional state field and conditions are because we can't stop timer started with `no_interval` option :stuck_out_tongue: 

Regarding `send_interval` using it might be also a little bit dangerous as it will send a message even if the previous one hasn't been processed yet. Maybe that's the reason there is no wrapper for it in Elixir.